torch/nn/functional.py

```diff
@@ -2984,3 +2987,28 @@ def fold(input, output_size, kernel_size, dilation=1, padding=0, stride=1):
         raise NotImplementedError("Input Error: Only 3D input Tensors are supported (got {}D)".format(input.dim()))
         ret = input  # TODO: remove when jit supports exception control flow
     return ret
+
+
+@weak_script
+def pad_circular(input, padding):
+    # type: (Tensor, List[int]) -> Tensor
+    """
+    Arguments
+        :param input: tensor of shape :math:`(N, C_{\text{in}}, H, [W, D]))`
+        :param padding: (tuple): m-elem tuple where m is the degree of convolution
+    Returns
+        :return: tensor of shape :math:`(N, C_{\text{in}}, [D + 2 * padding[0],
+                 H + 2 * padding[1]], W + 2 * padding[2]))`
+    """
```

Comment:
After reading this implementation, I realized (this was NOT in the original bootcamp task) that we can get error checking (which is conspicuously missing here) by folding `pad_circular` into `torch.nn.functional.pad`, which has a number of appropriate checks. It's probably a good idea to make sure these implementations line up anyway. Would you mind doing this? Thanks (and sorry for not realizing that was a better way to do it to start with.)

Reply:
I actually read through the implementation of pad(...). I think it is a good idea to combine those two functions. I'll work on it.

Reply:
Hi @ezyang, I looked into the implementation of pad() and it looks that the padding requirements are slightly different there. E.g. it requires that the length of pad is even: `assert len(pad) % 2 == 0, 'Padding length must be divisible by 2'`

Reply:
Do we  still want to fold `pad_circular` under `pad`, @ezyang ? 
For Conv3D padding can look something like this: 
```
m = nn.Conv3d(16, 33, (3, 5, 2), stride=(2, 1, 1), padding=(4, 2, 0))
```
This will fail the checks. It looks like in pad(...) function paddings are expected to have different format.

Reply:
F.pad wants its `pad` to be divisible by 2, to cover two sides of a dimension.
For example, (left_pad, right_pad), or (left_pad, right_pad, top_pad, bottom_pad), etc.

So, you should make `circular` a mode in `F.pad`, and in Conv*d, translate the padding to F.pad's equivalent.

For example, 

```
m = nn.Conv3d(16, 33, (3, 5, 2), stride=(2, 1, 1), padding=(4, 2, 0))
```
should end up calling (internally to Conv3d) `F.pad(input, (0, 0, 1, 1, 2, 2), mode='circular')`.

Note carefully that in Conv3d, the padding is given from left dimension to right dimension, i.e. `4` is the padding for D, `2` for H and `0` for W dimension of the input of dimensions N x C x D x H x W. And in `F.pad`, one gives padding dimension from right to left, so you give `(W_left, W_right, H_left, H_right, D_left, D_right)`.

Reply:
Hi @NarineK; sorry about the late response. It looks like `torch.nn.functional.pad` supports asymmetric padding (which is not directly supported by convolution). So, to convert a convolution style padding into a pad style padding, you would have to write `(4, 4, 2, 2, 0, 0)`.

Reply:
@ezyang see my comment above. it's not (4, 4, 2, 2, 0, 0), my comment gives the right ordering, reasoning and values.

Reply:
Somehow, your comment didn't show up at all when I was looking.

Reply:
I see, this will change the logic for the padding. In case of padding=(5, 2, 0)). I'll have something like: ```F.pad(input, (0, 0, 1, 1, 2, 3)```

Reply:
Pushed the changes

Reply:
addressed